README.md
# express-simple-logger

[![npm version](https://img.shields.io/npm/v/express-simple-logger.svg?style=flat)](https://www.npmjs.com/package/express-simple-logger)
[![build status](https://img.shields.io/travis/nicholascao/express-simple-logger.svg?style=flat)](https://travis-ci.org/NicholasCao/express-simple-logger)

 A logger middleware for [express](https://github.com/expressjs/express).Similar with [koa-logger](https://github.com/koajs/logger).

```
[2019-7-22 17:24:30] <-- GET /
[2019-7-22 17:24:31] --> GET / 200 835ms
[2019-7-22 17:24:48] <-- GET /
[2019-7-22 17:24:49] --> GET / 200 960ms
[2019-7-22 17:25:27] <-- GET /api
[2019-7-22 17:25:28] --> GET /api 200 357ms
[2019-7-22 17:25:53] <-- GET /users?page=2
[2019-7-22 17:25:54] --> GET /users?page=2 200 466ms
```

## Installation

Install via [npm](https://npmjs.com) or [yarn](https://yarnpkg.com)

```bash
# Use npm
$ npm i express-simple-logger

# Use yarn
$ yarn add express-simple-logger
```

## Example

```js
const express = require('express')
const logger = require('express-simple-logger')

const app = express()

app.use(logger())
```

## Notes

  Recommended that you `.use()` this middleware near the top
  to "wrap" all subsequent middleware.

## Option

|params|type|required|default|description|
|---|---|---|---|---|
|unless|Array|false|[]|ignore path list|
|logTime|Boolean|false|true|whether to log time|
|logger|Function|false|--|custom logger  (str:String, args:Array)|

  Param `str` is output string, args.join(' ')
  Param `args` is a array by `[logTime, format, method, url, status, time]`, when opt.logTime is false, logTime = ''

```js
const http = require('http')

...
app.use(logger({
  unless: ['/ignorepath'],
  logTime: false,
  // custom logger
  // add short description of status code
  logger (str, args) {
    if (args.length === 6) {
      // response logger
      args[4] = args[4] + ' ' + http.STATUS_CODES[args[4]]
      console.log(...args)
    }
  }
}))
```

## License

  MIT


CHANGELOG.md
## 2019-07-23
feature: add logTime option


test.js
'use strict'

// test tool
const chai = require('chai')
const request = require('supertest')
const sinon = require('sinon')
const sc = require('sinon-chai')
chai.use(sc)

const expect = chai.expect

let log, sandbox, app

/**
 * test cases
 */
describe('express-simple-logger', function () {
  before(function () {
    app = require('./test-server')()
  })

  beforeEach(function () {
    sandbox = sinon.createSandbox()
    log = sandbox.spy(console, 'log')
  })

  afterEach(function () {
    sandbox.restore()
  })

  it('should log a request', function (done) {
    request(app.listen()).get('/200').expect(200, 'hello world', function () {
      expect(log).to.have.been.calledTwice
      done()
    })
  })

  it('should log a request with correct method and url', function (done) {
    request(app.listen()).head('/200?logger=express-simple-logger').expect(200, function () {
      expect(log).to.have.been.calledWith(sinon.match.string, '<--', 'HEAD', '/200?logger=express-simple-logger')
      done()
    })
  })

  it('should log a response', function (done) {
    request(app.listen()).get('/200').expect(200, function () {
      expect(log).to.have.been.calledWith(sinon.match.string, '<--', 'GET' ,'/200')
      done()
    })
  })

  it('should log a 200 response', function (done) {
    request(app.listen()).get('/200').expect(200, function () {
      expect(log).to.have.been.calledWith(sinon.match.string, '-->', 'GET', '/200', 200, sinon.match.string)
      done()
    })
  })

  it('should log a 301 response', function (done) {
    request(app.listen()).get('/301').expect(301, function () {
      expect(log).to.have.been.calledWith(sinon.match.string, '-->', 'GET', '/301', 301, sinon.match.string)
      done()
    })
  })

  it('should log a 304 response', function (done) {
    request(app.listen()).get('/304').expect(304, function () {
      expect(log).to.have.been.calledWith(sinon.match.string, '-->', 'GET', '/304', 304, sinon.match.string)
      done()
    })
  })

  it('should log a 404 response', function (done) {
    request(app.listen()).get('/404').expect(404, function () {
      expect(log).to.have.been.calledWith(sinon.match.string, '-->', 'GET', '/404', 404, sinon.match.string)
      done()
    })
  })

  it('should log a 500 response', function (done) {
    request(app.listen()).get('/500').expect(500, function () {
      expect(log).to.have.been.calledWith(sinon.match.string, '-->', 'GET', '/500', 500, sinon.match.string)
      done()
    })
  })
})

describe('express-simple-logger with option', function () {
  before(function () {
    app = require('./test-server')({
      unless: ['/ignorepath'],
      logTime: false
    })
  })

  beforeEach(function () {
    sandbox = sinon.createSandbox()
    log = sandbox.spy(console, 'log')
  })

  afterEach(function () {
    sandbox.restore()
  })

  it('should log a response', function (done) {
    request(app.listen()).get('/200').expect(200, function () {
      expect(log).to.have.been.calledWith('', '<--', 'GET' ,'/200')
      expect(log).to.have.been.calledWith('', '-->', 'GET', '/200', 200, sinon.match.string)
      done()
    })
  })

  it('should not log', function (done) {
    request(app.listen()).get('/ignorepath').expect(200, 'ignorepath', function () {
      expect(log).to.have.callCount(0)
      done()
    })
  })
})

describe('express-simple-logger with custom logger', function () {
  let opt = {
    logger (str, args) {
      // do sth
    }
  }

  before(function () {
    app = require('./test-server')(opt)
  })

  beforeEach(function () {
    sandbox = sinon.createSandbox()
    log = sandbox.spy(opt, 'logger')
  })

  afterEach(function () {
    sandbox.restore()
  })

  it('should log with custom logger', function (done) {
    request(app.listen()).get('/200').expect(200, function () {
      let args = [sinon.match.string, '<--', 'GET' ,'/200']
      expect(log).to.have.been.calledWith(sinon.match.string, args)
      args = [sinon.match.string, '-->', 'GET' ,'/200', 200, sinon.match.string]
      expect(log).to.have.been.calledWith(sinon.match.string, args)
      done()
    })
  })
})


index.js
'use strict'

const dateFormat = function () {
  const padStart = function (num) {
    return String(num).padStart(2, '0')
  }

  let now = new Date()
  let year = now.getFullYear(),
    month = padStart(now.getMonth() + 1),
    day = padStart(now.getDay()),
    hour = padStart(now.getHours()),
    minute = padStart(now.getMinutes()),
    second = padStart(now.getSeconds())

  return `[${year}-${month}-${day} ${hour}:${minute}:${second}]`
}

const defaultOption = {
  unless: [],
  logTime: true
}

module.exports = function (opt = {}) {
  const log = function (...args) {
    if (opt.logger) {
      let str = args.join(' ')
      opt.logger(str, args)
    } else {
      console.log(...args)
    }
  }

  return function (req, res, next) {
    opt = { ...defaultOption, ...opt}

    if (opt.unless.indexOf(req.path) > -1) {
      return next()
    }

    let start = Date.now()

    log(opt.logTime ? dateFormat() : '', '<--', req.method, req.originalUrl)

    next()

    const onfinish = done.bind(null, 'finish')
    const onclose = done.bind(null, 'close')

    res.once('finish', onfinish)
    res.once('close', onclose)

    function done (event) {
      res.removeListener('finish', onfinish)
      res.removeListener('close', onclose)
      let upstream = event === 'close' ? '-x-' : '-->',
        delta = Date.now() - start,
        time = delta > 1000 ? Math.round(delta / 1000) + 's' : delta + 'ms'
      log(opt.logTime ? dateFormat() : '', upstream, req.method, req.originalUrl, res.statusCode || 404, time)
    }

  }
}


test-server.js
'use strict'

/**
 * test server
 */

const express = require('express')

const logger = require('./index')


module.exports = function (opt) {
  const app = express()
  app.use(logger(opt))
  
  app.use(function (req, res, next) {
    if (req.path === '/200') res.send('hello world')
    else return next()
  })
  
  app.use(function (req, res, next) {
    if (req.path === '/301') res.sendStatus(301)
    else return next()
  })
  
  app.use(function (req, res, next) {
    if (req.path === '/304') res.sendStatus(304)
    else return next()
  })
  
  app.use(function (req, res, next) {
    if (req.path === '/404') res.status(404).send('not found')
    else return next()
  })
  
  app.use(function (req, res, next) {
    if (req.path === '/500') res.status(500).send('server error')
    else return next()
  })
  
  app.use(function (req, res, next) {
    if (req.path === '/ignorepath') res.send('ignorepath')
    return next()
  })

  return app
}
